repo/blob/azure/azure_storage.go

```diff
@@ -275,18 +276,23 @@ func New(ctx context.Context, opt *Options, isCreate bool) (blob.Storage, error)
 			fmt.Sprintf("https://%s/", storageHostname), cred, nil,
 		)
 
+	// Azure AD
 	default:
-		return nil, errors.Errorf("either storage key or SAS token must be provided")
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
```

Comment:
@julio-lopez Here is the code block for supporting Azure AD

Reply:
@ywk253100 Can you add a test for this?

take a look at the "storage provider" test in this package for an example of how to pass credentials to tests. Now, in this case this is falling back to the SDK default behavior, so that may require a different set of tests.

Reply:
hi @ywk253100 and @julio-lopez , the usage of `NewDefaultAzureCredential` is fine in this case. However, I'd like to point out that there might be an implementation bug in this function:
`NewDefaultAzureCredential` checks through 4 different possible Az credentials type and keeps track of the results via an array of type `TokenCredential` (an interface) called `creds`. The returned credentials in each check is an implementation of interface `TokenCredential`.  If a credential check is successful `NewDefaultAzureCredential` append  the returned credential type. If the credential check failed, `NewDefaultAzureCredential` appends an error reporter which is also an implementation of interface `TokenCredential".  
At the end of function, `NewDefaultAzureCredential` returns error if either of this is true:
1.  len of `creds` is 0 (in `defaultAzureCredentialConstructorErrorHandler` and also `NewChainedTokenCredential`) 
2.  any of the element in `creds` is `nil` (`NewChainedTokenCredential`) 

IIUC, len of `creds` will always be 4. Even if none of the check succeed, `creds` array would still contain 4 error reoporter which are 4 implementation of `TokenCredential`. 
What that means for us here is that we no longer fail early here. If we would like to continue early failure, my suggestion is to add 
```go
// get NewDefaultAzureCredential(nil) cred
token, err:= cred.GetToken(ctx, nil)
// then check for error
```

Reply:
@leuyentran Thanks for the deep and thorough analysis.

Reply:
@ywk253100 you are right, the other two cases are not trying to get a token to validate the credentials. So, let's not add that here.

Having said that, @leuyentran point stands, the current implementation for `NewDefaultAzureCredential` does not return an error, even if there were errors attempting to obtain the different types of default credentials. It does not return an error and instead returns an error provider that will (lazily) fail later when `GetToken` is called.

Reply:
I opened an issue for Azure SDK: https://github.com/Azure/azure-sdk-for-go/issues/21301